EIP-1559 support in the proxy.

evgenyzdanovich opened the pull request:
Full support of EIP-1559 in the Proxy. Includes:

- New EIP-1559 RPC methods: eth_feeHistory, eth_maxPriorityFeePerGas
- Adjusted models: BlockHdrModel, TransactionModel, TransactionReceiptModel
- Adjusted existing RPC methods: eth_getBlockBy*, eth_getTransactionBy*
- DB changes to store the bits mentioned above

evgenyzdanovich committed:
Introduce DynamicGas transaction type in the Transaction and Transaction Model.

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -1,18 +1,20 @@
 from eth_hash.auto import keccak
 from typing_extensions import Self
 
+from common.utils.pydantic import BaseModel
```
Isn't used
Please remove it

evgenyzdanovich replied:
done

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -1,18 +1,20 @@
 from __future__ import annotations
 
-from typing import Final
+from typing import Final, List, Tuple, Union
```
Proxy is built for Ubuntu 22.04, which has Python 3.10 under the hood.
Since Python 3.9, it is preferred to use direct types: `list`, `tuple`, instead of `typing.List`, `typing.Tuple`

evgenyzdanovich replied:
done

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -97,10 +93,10 @@ def has_chain_id(self) -> bool:
 
     @staticmethod
-    def calc_chain_id(v: int) -> int | None:
+    def _calc_chain_id(v: int) -> int | None:
```
This method is used when loading transactions from DB.
But this can be refractory at the next steps.

evgenyzdanovich replied:
I can't see any references where it is used explicitly, I searched across the whole neon-proxy.py.

afalaleev replied:
There are two places:
- indexer/db/neon_tx_db.py:268 - neon_tx = NeonTxModel(...)
- indexer/base/objects.py:414 - return NeonTxModel(...)

In both places, there is no `chain-id` -property, but they have the `v`-property.
And you removed the cached-property `chain-id` from the NeonTxModel:
```
@cached_property		
def chain_id(self) -> int | None:		
    return EthTx.calc_chain_id(self.v)
```

So, before your modifications, the Proxy loaded NeonTxModel without `chain-id`, but with `v`.
And then calculated `chain-id` from the `v`.

Storing Tx as a blob-RLP requires additional steps because converting the old db to the new - is a very long process of about 0.5 hours.

So, we will have a grace period, when we should support 2 modes:
- if a record doesn't have a blob-RLP - load just `v` and convert it to `chain-id` for legacy txs
- if the record has a blob-RLP - loading blob-RLP and extracting all tx-properties from it

Or another variant:
- if the record doesn't have a blob-RLP - load parsed Tx, pack into RLP, and extract the result from RLP ))
- if the record has a blob-RLP - just load blob-RLP and extract all tx-properties from it

The second variant looks better because we can use the same logic for extracting tx-properties. And after converting DB in the background, we can remove the legacy code.

evgenyzdanovich replied:
Thank you! I'll restore the chain_id, my bad, not sure why I did that in the first place.

I'll go with the second option.

evgenyzdanovich replied:
@afalaleev do I understand correctly, that the plan is to have an additional blob column (let's call it "tx_rlp") and **leave other columns in place**? 
This way obviously our table will grow ~2 times, but the benefit is just convenient construction of `NeonTxMetaModel` in the code. 

Could you please confirm I understood it correctly?

evgenyzdanovich replied:
+CC @spylogsster, probably is more logical to do it as part of your changes.

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -183,3 +179,238 @@ def contract(self) -> bytes | None:
+    to_address: bytes
+    value: int
+    call_data: bytes
+    access_list: List[Tuple[int, List[int]]]
```
`access_list: list[tuple[int, list[int]]]`

afalaleev replied:
Why is this a list of ints?
As I know, it should list of `bytes` which then should be converted to list of `EthAddress`

evgenyzdanovich replied:
you are right. thanks, done

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -183,3 +179,238 @@ def contract(self) -> bytes | None:
+
+class EthTx:
+    type: int
+    payload: Union[EthLegacyTxPayload, EthDynamicGasTxPayload]
```
`EthLegacyTxPayload | EthDynamicGasTxPayload`

afalaleev replied:
IMHO, EthNoChainLegacyTxPayload should be part of this value instead of hiding it in EthLegacyTxPayload

evgenyzdanovich replied:
I agree, and this was my original intent tbh. 

But it has the same transaction type and it's harder to construct EthTx object, especially in the NeonTxModel -> EthTx creation (**NeonTxModel::_to_eth_tx method**): rlp objects - our transaction payloads - require the exact kwargs as defined by the `fields` field.

I'll leave it as is if you don't mind.

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -183,3 +179,238 @@ def contract(self) -> bytes | None:
+        if tx_type <= 0x7f:
+            # Typed transaction.
+            if tx_type not in (0, 2):
+                raise TypeError(f"Invalid transaction type parsed: {tx_type}")
```
Please, replace it with ValueError, because ValueError is handled by Pydantic.

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -1,6 +1,6 @@
 from __future__ import annotations
 
-from typing import Union
+from typing import Annotated, Literal, Optional, Union
```
Use `TYPE | None` instead of `typing.Optional`

afalaleev replied:
`Annotated`, `Literal` aren't used

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
+
+_TX_MODEL_EXCLUDE_LIST= {
+    0: set(["max_priority_fee_per_gas", "max_fee_per_gas", "chain_id"]),
+    2: set("gas_price")
```
`set()` gets the iterable object in the ctor.
And a string is an iterable object to, so it will be split into a list of characters.
As a result: 
`set("gas_price") -> {'g', 'a', 's', '_', 'p', 'r', 'i', 'c', 'e'}`

afalaleev replied:
set(["gas_price"])

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
 
 class NeonTxModel(BaseModel):
     tx_type: HexUIntField
+    chain_id: Optional[HexUIntField]
```
`HexUIntField | None`

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
     contract: EthAddressField
     nonce: HexUIntField
-    gas_price: HexUIntField
+    gas_price_legacy: Optional[HexUIntField] = Field(serialization_alias="gas_price")
```
`HexUIntField | None`

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
     nonce: HexUIntField
-    gas_price: HexUIntField
+    gas_price_legacy: Optional[HexUIntField] = Field(serialization_alias="gas_price")
+    max_priority_fee_per_gas: Optional[HexUIntField]
```
`HexUIntField | None`

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
-    gas_price: HexUIntField
+    gas_price_legacy: Optional[HexUIntField] = Field(serialization_alias="gas_price")
+    max_priority_fee_per_gas: Optional[HexUIntField]
+    max_fee_per_gas: Optional[HexUIntField]
```
`HexUIntField | None`

evgenyzdanovich replied:
done

afalaleev commented on tests/neon/test_transaction.py:
```diff
@@ -3,9 +3,16 @@
 class TestNeonTx(unittest.TestCase):
-    raw_tx = "0xF86B80850BA43B7400825208947917bc33eea648809c285607579c9919fb864f8f8703BAF82D03A0008025A0067940651530790861714b2e8fd8b080361d1ada048189000c07a66848afde46A069b041db7c29dbcc6becf42017ca7ac086b12bd53ec8ee494596f790fb6a0a69"
+    # https://etherscan.io/tx/0x14a298c1eea89f42285948b7d51eeac2876ca7406c9784b9b90dd3591d156d64
+    raw_tx = '0xF86B80850BA43B7400825208947917bc33eea648809c285607579c9919fb864f8f8703BAF82D03A0008025A0067940651530790861714b2e8fd8b080361d1ada048189000c07a66848afde46A069b041db7c29dbcc6becf42017ca7ac086b12bd53ec8ee494596f790fb6a0a69'
```
Please use the `black`-utility with `-l120` for code formatting.

evgenyzdanovich replied:
done, but it didn't make changes, only minor reformatting - changing strings to double quotes, and other stuff :(

I took a look at their issues and found some. Reported I have this problem: https://github.com/psf/black/issues/1331

In the meantime, I split it manually.

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -183,3 +179,238 @@ def contract(self) -> bytes | None:
+            elif tx_type == 2:
+                payload_cls = EthDynamicGasTxPayload
+            else:
+                raise TypeError(f"Invalid transaction type specified: {tx_type}")
```
Please, replace it with `ValueError`, because `ValueError` is handled by Pydantic.

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
 from ..utils.format import str_fmt_object
 from ..utils.pydantic import BaseModel, HexUIntField
 
+from pydantic import Field, PlainSerializer, PlainValidator, model_validator
```
`PlainSerializer`, `PlainValidator` aren't used.

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -11,34 +11,62 @@
     error: str | None = None
 
+    @model_validator(mode="after")
+    def check_gas_price_fields_and_transaction_type(self) -> Self:
```
Please use `def model_post_init(self, _ctx) -> None:` instead

evgenyzdanovich replied:
done

evgenyzdanovich committed:
Adjust according to the code review and fix the test.

afalaleev commented on common/ethereum/transaction.py:
```diff
@@ -12,7 +12,7 @@
 from ..utils.format import hex_to_bytes
 
```
Please run `black` utility to format the source code.

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -1,44 +1,70 @@
 from __future__ import annotations
 
```
Please run `black`-utility on the source code

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -1,44 +1,70 @@
+_TX_MODEL_EXCLUDE_LIST= {
+    0: set(["max_priority_fee_per_gas", "max_fee_per_gas", "chain_id"]),
+    2: set(["gas_price"])
+}
```
The recommendation from the profiler
```
_TX_MODEL_EXCLUDE_LIST = {
    0: {"max_priority_fee_per_gas", "max_fee_per_gas", "chain_id"},
    2: {"gas_price"},
}
```

evgenyzdanovich replied:
done

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -1,44 +1,70 @@
+                raise ValueError("gas_price shouldn't be specified for the Dynamic Gas transaction.")
+            if self.max_priority_fee_per_gas is None or self.max_fee_per_gas is None:
+                raise ValueError("max_priority_fee_per_gas or max_fee_per_gas is absent for the Dynamic Gas transaction.")
+        return self
```
remove 
`return self`
because the function doesn't return a value:
`def model_post_init(self, _ctx: Any) -> None:`

evgenyzdanovich replied:
done

evgenyzdanovich committed:
Second round of review.

afalaleev commented on indexer/base/objects.py:
```diff
@@ -410,6 +410,7 @@ class _DeprecatedNeonTx(BaseModel):
 
         def to_clean_copy(self) -> NeonTxModel:
+            # TODO EIP1559: adjust constructor params for DynamicGas transaction type.
             return NeonTxModel(
```
In addition to TODO.
`gas_price` is a property now, you should rename the `gas_price` to `gas_price_legacy` here

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -264,6 +264,7 @@ def to_meta(self) -> NeonTxMetaModel | None:
             return None
 
+        # TODO EIP1559: introduce blob field which stores rlp and construct via from_raw(rlp).
         neon_tx = NeonTxModel(
```
In addition to TODO.
`gas_price` is a property now, you should rename the `gas_price` to `gas_price_legacy` here

evgenyzdanovich replied:
done

evgenyzdanovich committed:
Another round of review.

evgenyzdanovich committed:
Temporary fix to set tx_chain_id as None for objects returned from the db.

afalaleev commented on common/neon/transaction_model.py:
```diff
@@ -104,24 +139,48 @@ def _from_eth_tx(cls, tx: EthTx) -> Self:
+            s=self.s,
+        )
+        for pop_field in _TX_MODEL_EXCLUDE_LIST[self.tx_type]:
+            ctr.pop(pop_field)
```
The reason, why tests don't work:
- `TX_MODEL_EXCLUDE_LIST` contains `tx_chain_id`
- But this structure contains the `chain_id` property

evgenyzdanovich replied:
Amazing, thank you! basic-tests are passing now! 
dapps are still failing, I'll commit general transaction processing logic in 1-2 hours, let's see how it goes.

evgenyzdanovich committed:
Fix the test.

evgenyzdanovich committed:
General transaction processing logic (for dynamic gas transactions): add sysvar account, set corresponding compute units price.

evgenyzdanovich committed:
Adjust getTransactionBy* response struct to support new fields from the Dynamic Gas transaction.

evgenyzdanovich committed:
Adjust a comment to change gasPrice returned in eth_getTransactionBy* to make it precise.

afalaleev commented on common/neon/neon_program.py:
```diff
@@ -87,15 +88,17 @@ class NeonProg:
     _protocol_version: ClassVar[NeonEvmProtocol] = NeonEvmProtocol.v1011
     _evm_version: ClassVar[str] = "v1.11.x"
     ID: ClassVar[SolPubKey] = NEON_EVM_PROGRAM_ID
+    IX_SYSVAR_ID: ClassVar[SolPubKey] = INSTRUCTIONS_SYSVAR_ACCOUNT
```
Please move the variable to `SolSysProg`.

evgenyzdanovich replied:
done.

afalaleev commented on common/neon/neon_program.py:
```diff
@@ -87,15 +88,17 @@ class NeonProg:
     # +6: NeonProg.ID
     # +7: CbProg.ID
-    BaseAccountCnt: Final[int] = 7
+    # +8: Instructions Sysvar account
```
The logic of this comment:
- positions 1-5 (without prefix `+`) - are required accounts in a Neon instruction
- positions 6-7 (with the prefix `+`) - aren't listed in the Neon instruction, BUT! they are added when the Proxy creates the Solana transactions and they are listed in the program-ids of instructions.

So, to make the comment correct, you should place the Sysvar account on the position 6 without `+`, because it is a required account for a Neon instruction.
NeonProg.ID and CbProg.ID should be changed to +7, +8:
```
1. holder
2. payer
3. treasury-pool-address,
4. payer-token-address
5. SolSysProg.ID
6. Instructions Sysvar account
+7. NeonProg.ID
+8. CbProg.ID
```
IMHO, it will be great to add the comment about the logic, but I don't require it from you.

evgenyzdanovich replied:
thanks, you solved the mystery for me :) adjusted.

afalaleev commented on proxy/executor/strategy_base.py:
```diff
@@ -113,7 +113,16 @@ async def cancel(self) -> ExecTxRespCode | None:
 
     @property
     def _cu_price(self) -> int:
```
strategy_iterative.py has its own logic, please fix it too.

evgenyzdanovich replied:
removed the override.

afalaleev commented on proxy/executor/strategy_base.py:
```diff
@@ -113,7 +113,16 @@ async def cancel(self) -> ExecTxRespCode | None:
+        assert base_fee_per_gas > 0
+        max_priority_fee_per_gas = self._ctx.max_priority_fee_per_gas
+        # TODO EIP1559: conversion to micro lamports???
+        return max(1, int(max_priority_fee_per_gas * 5000.0 / (base_fee_per_gas * self._cu_limit)))
```
Yep, here should be the multiplication on 1'000'000.

In the bottom part of the formula, you have cu_limit which is equal to 1'400'000, and after direct division, you will get the number less than 1.

evgenyzdanovich replied:
thanks, done

afalaleev reviewed (commented):
Looks good!
Let's finish it.

evgenyzdanovich committed:
Code review changes.

evgenyzdanovich committed:
Introduce chain_id, max_fee_per_gas, max_priority_fee_per_gas in the NeonTxDb table.

evgenyzdanovich committed:
Tiny black util reformatting.

evgenyzdanovich committed:
Implement priority fee log decoder, accumulate it across neon iterations and store priority fee in the Receipt Model and in the NeonTxDb.

evgenyzdanovich committed:
Minor fix of ReceiptModel construction from deprecated one.

evgenyzdanovich committed:
Calculate the effectiveGasPrice for the new transaction type in the receipt RPC.

evgenyzdanovich committed:
Adjust eth_getBlockBy* to include baseFeePerGas into the response.

evgenyzdanovich committed:
Introduce new RPC methods: eth_feeHistory and eth_getMaxPriorityFeePerGas.

Also, related changes to:
- DB, solana_block table now includes priority fee percentiles.
- A corresponding BlockHdrModel now also includes percentiles.
- A new config value that denotes the number of blocks to take the average priority fee from.

evgenyzdanovich committed:
Resolve TODOs, part 1: effective gas price.

evgenyzdanovich committed:
Resolve TODOs, part2: wrap neon transaction type into Enum, expose helpers and use it.

evgenyzdanovich committed:
Tiny fix: eth_feeHistory.

evgenyzdanovich committed:
Small bug fixes while looking at new tests here and there.

evgenyzdanovich committed:
Another small fixes: make sure NeonTxModel is constructed correctly.

evgenyzdanovich committed:
Fix NeonTxModel -> EthTx - add the missing access_list.

evgenyzdanovich committed:
Fix feeHistory RPC: typo in the SQL query and more graceful handling of input params and special cases.

evgenyzdanovich committed:
Fix conversion to BIGINT in SQL: integer as a hex text can't be directly converted to BIGINT.

evgenyzdanovich committed:
Fix NeonTxModel serialization <=> deserialization.

evgenyzdanovich committed:
Fix of typo in BlockFeeGasData

evgenyzdanovich committed:
Another small fixes.

evgenyzdanovich committed:
Add validation on gas pricing in the neon transaction model.

afalaleev commented on common/config/config.py:
```diff
@@ -169,7 +170,7 @@ class Config:
     fuzz_fail_pct_name: Final[str] = "FUZZ_FAIL_PCT"
 
-    _pg_null_value: Final[object] = object()
+    _pg_null_value: Final[str] = ""  # to have a uniform (str) type returned by pg_* properties.
```
It is a Python trick that allows to define an empty parameter (for example, the empty password), but simultaneously checks that the parameter is defined.

By default, all Postgres parameters are `_pg_null_value`, so if a parameter isn't defined in the config, the value will be a `_pg_null_value`.

DbConnection in ctor calls `config.validate_db_config() `, which checks the parameters are defined, the statement ` v is v1` in Python checks that it is the same object (not! value).

But in Python all empty strings are the same, you can check it by the following examples:
```
v = ""
v1 = v
print(v1 is v)

v1 = ""
print(v1 is v)
```
You will see:
```
True
True
```

At the same time:
```
v = object()
v1 = v
print(v1 is v)

v1 = object()
print(v1 is v)
```
You will see:
```
True
False
```

evgenyzdanovich replied:
1. Sorry, I don't understand how it makes logic different from assigning it to empty string? We currently still return `self._pg_null_value` from pg_* properties (the same exact object) and compare it to `self._pg_null_value`, so in both cases (whether it's an empty string or the same exact `object()`) it seems to work the same.  
2. Also, the only reason I made an adjustment is because test_hide_sensitive fails:
"TypeError: object of type 'object' has no len()", thrown from `sensitive_info_list property`
3. Lastly, the pg_* properties returns str and `config/utils.py` seems to handle only str and list and is not aware that that default for pg_* properties can be object.

What should I do with it? I can try to reconcile it, if you want me to do it.

evgenyzdanovich replied:
Discussed f2f - reverted these changes. Fixed the unit test (`test_hide_sensitive`) with the default parameters when invoked directly (and not through the `deploy_check`)

afalaleev commented on common/neon/block.py:
```diff
@@ -19,6 +22,7 @@ class NeonBlockHdrModel(BaseModel):
     block_time: int | None
     parent_slot: int | None
     parent_block_hash: EthBlockHashField
+    priority_fee_percentiles: list[int]
```
Please rename to:
`cu_price_percentile_list`

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -65,8 +71,25 @@ def _from_sol_block(cls, raw: SolRpcBlockInfo) -> Self:
+            return [0] * 11
+
+        # Build a full list of compute unit prices in the solana block.
+        prices: list[int] = list()
```
price_list

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -1,9 +1,12 @@
 
 from typing_extensions import Self
 
+from common.solana.transaction_decoder import SolTxMetaInfo
```
`from ..solana.transaction_decoder import SolTxMetaInfo`

afalaleev replied:
Does your IDE support optimizing imports?

evgenyzdanovich replied:
Adjusted the corresponding setting in my VS Code

Done the change.

afalaleev commented on common/neon/transaction_meta_model.py:
```diff
@@ -13,3 +13,14 @@ class NeonTxMetaModel(BaseModel):
     def neon_tx_hash(self) -> EthTxHash:
         return self.neon_tx.neon_tx_hash
+
+    @property
```
`@cached_property`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -324,6 +382,7 @@ def to_meta(self) -> NeonTxMetaModel | None:
             status=self.status,
             total_gas_used=self.gas_used,
             sum_gas_used=self.sum_gas_used,
+            priority_fee_spent=self.priority_fee_spent if self.priority_fee_spent else 0,
```
`priority_fee_spent=self.priority_fee_spent or 0`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -96,16 +97,46 @@ async def start(self) -> None:
+            chain_id=DbSqlParam("chain_id"),
+            num_blocks=DbSqlParam("num_blocks"),
+            latest_slot=DbSqlParam("latest_slot"),
+        )
```
This SQL is very slow in big db.
It should be refactored.

evgenyzdanovich replied:
Agree, I'm thinking about it, if you have concrete ideas, please let me know.

evgenyzdanovich replied:
Discussed f2f: the main problem you spotted is that the inner select is not bounded by the amount entries it returns (and will return a lot of entries).

So, I moved `LIMIT` and `WHERE` conditional into the inner. It should work fast right now and quite accurate.

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -29,6 +29,7 @@ def __init__(self, db: DbConnection):
         self._select_by_nonce_query = DbQueryBody()
         self._select_by_block_query = DbQueryBody()
         self._select_by_index_query = DbQueryBody()
+        self._select_base_fees_query = DbQueryBody()
```
Please rename to
`self._select_base_fee_list_query`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -96,16 +97,46 @@ async def start(self) -> None:
             index=DbSqlParam("index"),
         )
 
+        select_base_fees_sql = DbSql(
```
Please rename to:
`select_base_fee_list_sql`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -156,6 +187,16 @@ async def get_tx_by_slot_tx_idx(self, ctx: DbTxCtx, slot: int, tx_idx: int) -> N
         )
         return _RecordWithBlock.to_meta(rec)
 
+    async def get_base_fees(
```
Please rename to
`get_base_fee_list`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/solana_block_db.py:
```diff
@@ -177,20 +178,37 @@ async def start(self) -> None:
             slot_list=DbSqlParam("slot_list"),
         )
 
+        priority_fee_percentiles_sql = DbSql(
```
Please rename to:
`select_cu_price_percentile_list_sql`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/solana_block_db.py:
```diff
@@ -34,6 +34,7 @@ def __init__(self, db: DbConnection):
         self._finalize_query = DbQueryBody()
         self._deactivate_query = DbQueryBody()
         self._activate_query = DbQueryBody()
+        self._priority_fee_percentiles_query = DbQueryBody()
```
Please rename to:
`self._select_cu_price_percentile_list_query`

evgenyzdanovich replied:
done

afalaleev commented on proxy/executor/strategy_base.py:
```diff
@@ -113,7 +113,15 @@ async def cancel(self) -> ExecTxRespCode | None:
 
     @property
```
`@cached_property`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -16,6 +20,10 @@
 
 
+# Hardcoded priority fee percentiles that are supported by eth_feeHistory.
+_REWARD_PERCENTILES: list[int] = [i * 10 for i in range(11)]
```
Please rename to:
`_REWARD_PERCENTILE_LIST`

afalaleev replied:
Numbers `10` and `11` should be defined in NeonBlockHdr

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -104,6 +112,26 @@ def model_post_init(self, _ctx) -> None:
 
+class _RpcFeeHistoryResp(BaseJsonRpcModel):
+    baseFeePerGas: list[HexUIntField]
+    gasUsedRatio: list[float]
```
Are there any requirements for precision in Ethereum protocol?

evgenyzdanovich replied:
not that I'm aware of. 
Also, it seems, Infura and Alchemy return arbitrary precision for floats.

BoogVAr replied:
Maybe it's better to return Decimal to have no surprise?

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        # Fetch the compute units price across last several blocks (as specified in the cfg).
+        num_blocks = self._cfg.priority_fee_num_blocks_to_average
+        assert num_blocks > 0
+        historical_priority_fees: list[list[int]] = await self._db.get_historical_priority_fees(num_blocks)
```
Please rename to:
`historical_cu_price_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        num_blocks = self._cfg.priority_fee_num_blocks_to_average
+        assert num_blocks > 0
+        historical_priority_fees: list[list[int]] = await self._db.get_historical_priority_fees(num_blocks)
+        assert num_blocks == len(historical_priority_fees)
```
`block_cnt =`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+
+        # Take the weighted average across fetched blocks.
+        average_weighted_max_priority_fee = 0
+        for idx, block_priority_fees in enumerate(historical_priority_fees):
```
Please rename to:
`block_cu_price_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        ctx: HttpRequestCtx,
+        num_blocks: HexUIntField,
+        block_tag: RpcBlockRequest,
+        priority_fee_percentiles: list[int] | None,
```
Please rename to:
`priority_fee_percentile_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+    async def get_fee_history(
+        self,
+        ctx: HttpRequestCtx,
+        num_blocks: HexUIntField,
```
Please rename to:
`block_cnt`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        # Infura and Alchemy currently do the same.
+        num_blocks = max(1, min(num_blocks, 1024))
+        # Fetching the data stage.
+        fee_gas_data: list[BlockFeeGasData] = await self._db.get_historical_base_fees(
```
Please rename to:
`fee_gas_data_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+                for p in percentiles
+            ]
+
+        rewards: list[list[int]] = [
```
Please rename to:
`reward_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/server_abc.py:
```diff
@@ -354,6 +354,7 @@ async def _init_genesis_block(self) -> None:
             block_time=block_time,
             parent_slot=0,
             parent_block_hash=parent_hash,
+            priority_fee_percentiles=[0] * 11,
```
Please rename to:
`priority_fee_percentile_list`

afalaleev replied:
Please move the magic constant to NeonBlockHdr

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        ]
+
+        # OldestBlock:
+        oldest_block = fee_gas_data[-1].block_slot if len(fee_gas_data) > 0 else 0
```
This simplifying is enough:
`if fee_gas_data_list else 0`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        base_fees.append(current_gas_price)
+
+        # GasUsedRatio:
+        gas_used_ratio = [
```
Please rename to:
`gas_used_ratio_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+
+        # Processing stage. Data entries are sorted in descending order by slot.
+        # BaseFees:
+        base_fees: list[int] = [
```
Please rename to:
`base_fee_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+
+        # Since we are effectively querying different tables, let's reconcile length of the data,
+        # so we return consistent data (in case one array is bigger than the other).
+        num_blocks = min(len(fee_gas_data), len(priority_fees_data))
```
Please rename to:
`block_cnt`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        fee_gas_data: list[BlockFeeGasData] = await self._db.get_historical_base_fees(
+            self.get_chain_id(ctx), num_blocks, latest_slot
+        )
+        priority_fees_data: list[list[int]] = await self._db.get_historical_priority_fees(num_blocks)
```
Please rename to:
`cu_price_data_list`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        for idx, block_priority_fees in enumerate(historical_priority_fees):
+            # Sanity check to avoid division by zero.
+            average_block_priority_fee = 0
+            if len(block_priority_fees) != 0:
```
Simple version is enough:
`if block_cu_price_list:`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+        average_weighted_max_priority_fee = 0
+        for idx, block_priority_fees in enumerate(historical_priority_fees):
+            # Sanity check to avoid division by zero.
+            average_block_priority_fee = 0
```
Please rename to:
`avg_block_cu_price`

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +168,122 @@ async def get_neon_gas_price(
+    @NeonProxyApi.method(name="eth_maxPriorityFeePerGas")
+    async def get_max_priority_fee_per_gas(self, ctx: HttpRequestCtx) -> HexUIntField:
+        # Fetch the compute units price across last several blocks (as specified in the cfg).
+        num_blocks = self._cfg.priority_fee_num_blocks_to_average
```
Please rename to:
`block_cnt`

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -264,13 +309,19 @@ def from_tx(cls, neon_tx: NeonTxModel, neon_rcpt: NeonTxReceiptModel) -> Self:
             value=hex(neon_tx.value),
             calldata=neon_tx.call_data.to_string(),
-            gas_price=hex(neon_tx.gas_price),
+            gas_price=hex(NeonTxMetaModel(neon_tx=neon_tx, neon_tx_rcpt=neon_rcpt).effective_gas_price),
```
This method is called from one place, and in the source line there is `NeonTxMetaModel` object.
Maybe it is better to pass the base object, instead of constructing it from the params?

evgenyzdanovich replied:
I'm not sure I understand it. Yes, `from_tx` is only called from `set_block_list`, but there's no NeonTxMetaModel.

I created an additional `@classmethod` that allows to do the same logic but without constructing class object. The `@cached_property` is also in place to support access to `effective_gas_price` from `np_block_transaction_api`.

WDYT?

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -324,6 +382,7 @@ def to_meta(self) -> NeonTxMetaModel | None:
             status=self.status,
             total_gas_used=self.gas_used,
             sum_gas_used=self.sum_gas_used,
+            priority_fee_spent=self.priority_fee_spent if self.priority_fee_spent else 0,
```
`priority_fee_spent=self.priority_fee_spent or 0`

evgenyzdanovich replied:
done

afalaleev commented on db/scheme.sql:
```diff
@@ -71,8 +71,11 @@ CREATE TABLE IF NOT EXISTS solana_blocks (
     is_finalized BOOL,
-    is_active BOOL
+    is_active BOOL,
+    priority_fee_percentiles INT ARRAY
```
Technically you store `compute unit prices` not `priority fees`.

Please rename to:
`cu_price_percentiles`

evgenyzdanovich replied:
done

afalaleev commented on common/config/config.py:
```diff
@@ -133,6 +133,7 @@ class Config:
     min_gas_price_name: Final[str] = "MINIMAL_GAS_PRICE"
     min_wo_chain_id_gas_price_name: Final[str] = "MINIMAL_WITHOUT_CHAIN_ID_GAS_PRICE"
     const_gas_price_name: Final[str] = "CONST_GAS_PRICE"
+    priority_fee_num_blocks_to_average_name: Final[str] = "PRIORITY_FEE_NUM_BLOCKS_TO_AVERAGE"
```
Please rename to:
`priority_fee_block_cnt_to_avg_name`

evgenyzdanovich replied:
done

afalaleev commented on common/config/config.py:
```diff
@@ -667,6 +668,10 @@ def const_gas_price(self) -> int | None:
         return const_gas_price * (10**9)
 
+    @cached_property
+    def priority_fee_num_blocks_to_average(self) -> int:
```
Please rename to:
`priority_fee_block_cnt_to_avg`

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -29,6 +33,7 @@ def default(cls) -> Self:
             block_time=None,
             parent_slot=None,
             parent_block_hash=EthBlockHashField.default(),
+            priority_fee_percentiles=[0] * 11,  # Although this path is not used, let's make it future-proof.
```
Please add the constant to the class definition instead of magic number:
 ```
 class NeonBlockHdrModel(BaseModel):
    PercentileCnt: Final[int] = 11
    ....
```

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -40,6 +45,7 @@ def new_empty(cls, slot: int, commit: EthCommit = EthCommit.Pending) -> Self:
             block_time=None,
             parent_slot=None,
             parent_block_hash=EthBlockHashField.default(),
+            priority_fee_percentiles=[0] * 11,  # Although this path is not used, let's make it future-proof.
```
Magic number ....

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -65,8 +71,25 @@ def _from_sol_block(cls, raw: SolRpcBlockInfo) -> Self:
+    def _calculate_priority_fee_stats(cls, sol_block: SolRpcBlockInfo) -> list[int]:
+        # In case for some reason, Solana block does not have any transactions.
+        if not sol_block.tx_list:
+            return [0] * 11
```
Magic number...

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -65,8 +71,25 @@ def _from_sol_block(cls, raw: SolRpcBlockInfo) -> Self:
+        # Sort.
+        prices.sort()
+        # Take every i * 10 percentile (i:=0..10) in a sorted list.
+        return [prices[math.ceil(p * (len(prices) - 1) / 100)] for p in range(0, 101, 10)]
```
Magic numbers...

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/solana_block_db.py:
```diff
@@ -233,7 +251,10 @@ async def _generate_fake_block(self, ctx: DbTxCtx, slot: int | None, slot_range:
-            return NeonBlockHdrModel(slot=slot)
+            return NeonBlockHdrModel(
+                slot=slot,
+                priority_fee_percentiles=[0] * 11,
```
Magic number...

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/solana_block_db.py:
```diff
@@ -243,8 +264,9 @@ async def _generate_fake_block(self, ctx: DbTxCtx, slot: int | None, slot_range:
-            parent_slot=slot-1,
+            parent_slot=slot - 1,
             parent_block_hash=self._generate_fake_block_hash(slot - 1),
+            priority_fee_percentiles=[0] * 11,
```
Magic number ...

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/solana_block_db.py:
```diff
@@ -312,6 +335,12 @@ async def activate_block_list(self, ctx: DbTxCtx, from_slot: int, slot_list: tup
         await self._update_row(ctx, self._deactivate_query, by_slot_range)
         await self._update_row(ctx, self._activate_query, by_slot_range)
 
+    async def get_priority_fee_percentiles(self, ctx: DbTxCtx, num_blocks: int) -> list[list[int]]:
```
Please rename to:
`get_cu_price_percentile_list`

evgenyzdanovich replied:
done

afalaleev commented on common/neon/block.py:
```diff
@@ -65,8 +71,25 @@ def _from_sol_block(cls, raw: SolRpcBlockInfo) -> Self:
         )
 
+    @classmethod
+    def _calculate_priority_fee_stats(cls, sol_block: SolRpcBlockInfo) -> list[int]:
```
Please rename to:
`_calc_cu_price_stat`

evgenyzdanovich replied:
done

evgenyzdanovich committed:
Code review changes.

evgenyzdanovich committed:
Fix sedes for Access List in the Dynamic Gas transaction.

evgenyzdanovich committed:
Catch validation error in the eth_sendRawTransaction and rethrow it as EthError to report proper error code.

evgenyzdanovich committed:
Small fixes after the rebase.

evgenyzdanovich committed:
Add validation in the NeonTxModel that chain_id should be specified for Dynamic Gas transaction.

evgenyzdanovich committed:
Populate the default chain_id for methods that return incomplete NeonTxModel (from_tx_hash, default, etc) - to comply with validation.

evgenyzdanovich committed:
Treat chain_id=0 separately from chain_id=None in the transaction validation.

evgenyzdanovich committed:
Add validation on eth_feeHistory percentiles.

evgenyzdanovich committed:
Fix test_hide_sensitive with the default parameters.

evgenyzdanovich committed:
Adjust eth_feeHistory to return zero base_fee_per_gas for the empty blocks.

evgenyzdanovich committed:
Remove deleted deprecated code that appeared after rebase by mistake.

evgenyzdanovich commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +175,169 @@ async def get_neon_gas_price(
+        fee_data_it: int = len(fee_gas_data_list) - 1
+        cu_data_it: int = len(cu_price_data_list) - 1
+        last_base_fee_per_gas: int = 0
+        while cu_data_it >= 0:
```
@afalaleev The logic here is the following:

we get transactions of the new type together with the `block_slot` (gas_fee_data).
we get priority fee percentiles of the solana (from all transactions) together with the `block_slot` - (cu_price_data).

Now, we "merge" those entries by the `block_slot`, so we have the consistent response (base_fee and reward correspond to each other, as they are from the same solana-neon blocks). We fill with zeroes (or with previous base_fee) for solana blocks that has empty neon block.

The problem I see is the following: most of the time, all the entries from `gas_fee_data` will be with the `block_slot` smaller than the ones from `cu_price_data` (because of lots of empty neon blocks). 
As a result, we return zeroes as the base_fee in the response.

As an idea: maybe it makes sense to not merge by the `block_slot` and treat them separately (construct base_fee in the response by relying **solely** on the `gas_fee_data`; construct `reward` in the response by relying **solely** on the `cu_price_data`? This way, we lose consistency of the response, but in contrast the response makes more sense (without zeroes, with **actual** base_fee from neon blocks, and with rewards from solana blockchain). WDYT?

evgenyzdanovich replied:
Implemented as discussed f2f: expose recent token gas price from the mempool and use it in the eth_feehistory in case User asks for the feehistory about recent blocks.
In case User requests historical data - take the data from the DB from the transactions as it was already implemented.

evgenyzdanovich committed:
Fix incorrect typing error in the HolderAccountModel.

evgenyzdanovich committed:
Rework eth_feehistory: expose recent token gas price from the mempool, use it in the eth_feehistory for the recent blocks.

evgenyzdanovich committed:
Make sure eth_feeHistory does not return blocks earlier than earliest neon block.

afalaleev commented on common/neon/neon_program.py:
```diff
@@ -95,14 +95,19 @@ class NeonProg:
-    # +6: NeonProg.ID
-    # +7: CbProg.ID
-    BaseAccountCnt: Final[int] = 7
+    # 6: SolSysProg.INSTRUCTIONS_SYSVAR
```
Just to comment on the already discussed topic.

It is a bad idea to add an account to the required list because the limit on the number of available accounts is critical.

There are two options:
- Add special instructions for typed transactions
- Check the possibility to use a syscall

evgenyzdanovich replied:
done, now it works without additional account - utilizes syscall.

afalaleev commented on indexer/base/objects.py:
```diff
@@ -358,6 +359,7 @@ class InitData(BaseNeonIndexedObjInfo.InitData):
         operator: SolPubKeyField
         gas_used: int
         total_gas_used: int
+        total_priority_fee: int
```
Define a default value for the total_priority_fee:
```
total_priority_fee: int = 0
```

The reason: on the upgrade there will be some stuck transactions with undefined priority fee.

evgenyzdanovich replied:
done

evgenyzdanovich committed:
Remove irrelevant todo.

evgenyzdanovich committed:
Filter out Vote transactions from cu-price stats calculation in the indexer.

evgenyzdanovich committed:
Rework eth_getBlockBy: for the recent blocks, return baseFeePerGas from the gas token price from the operator's mempool. Refactor cu_price estimation - do not rely on atlas, instead take data from indexer's cu price percentiles in the solana blocks.

evgenyzdanovich committed:
Slight refactoring: factor out some logic related to MpRecentGasPricesModel into the class itself.

evgenyzdanovich committed:
Get rid of Sysvar Account in the account list - evm is already capable of priority fee validation via syscalls.

evgenyzdanovich committed:
Account for cu_limit in the cu_price calculation formula for the new transaction type.

evgenyzdanovich committed:
Minor fix: construct CoreApiTxModel correctly based on transaction type.

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   58829 |      0     |      0.106 |    0.00132496  |         0     |
| getSlot                           |   57831 |      0     |      0.165 |    0.000796545 |         0     |
| getBlocks                         |   22582 |      0     |      0.043 |    0.000636348 |         0     |
| getBlock                          |   14610 |      0     |      0.049 |    0.00134209  |         0.001 |
| getMultipleAccounts               |   13298 |      0     |      0.082 |    0.00175395  |         0.001 |
| getTransaction                    |    8130 |      0     |      0.212 |    0.00826458  |         0.003 |
| getBlockTime                      |    6215 |      0     |      0.043 |    0.000455189 |         0     |
| sendTransaction                   |    5252 |      0     |      0.108 |    0.00634044  |         0.003 |
| getLatestBlockhash                |    5073 |      0     |      0.092 |    0.00213424  |         0.001 |
| getBalance                        |    1846 |      0     |      0.044 |    0.00115493  |         0     |
| getSignatureStatuses              |    1152 |      0     |      0.036 |    0.00101128  |         0     |
| getVersion                        |     600 |      0     |      0.034 |    0.000975    |         0     |
| isBlockhashValid                  |     558 |      0     |      0.022 |    0.00040681  |         0     |
| getSignaturesForAddress           |      62 |      0     |      0.004 |    0.00108065  |         0.001 |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000342105 |         0     |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000625    |         0.001 |
| getHealth                         |       3 |      0.001 |      0.027 |    0.011       |         0.005 |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000666667 |         0.001 |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |         0     |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

</details>

evgenyzdanovich committed:
Cover the case when max_fee_per_gas = max_priority_fee_per_gas - treat it as a legacy transaction.

evgenyzdanovich commented:
@afalaleev please take another look.

since the last time you had a review - here's the list of what we agreed on and what was changed:

- Adjusted various RPCs to return baseFeePerGas from the operators mempool pricing. It includes getGasPrice, getBlockBy, etc
- Reworked and exposed indexer based minimalistic estimator of priority fees.
 It takes all the solana transactions that come to indexer, removes `vote1111` from it and calculates 10th percentiles. Later, it's used in the cu_price estimation and in RPC  getMaxPriorityFeePerGas (basic interpolation between percentiles).
- Removed sysvar account for DynamicGas transactions - substituted with syscall.

P.S. Currently, the usage of Atlas is commented out. I suggest to clean it up once we confirm the new mechanism works correctly in dev/main nets.

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   59409 |      0     |      0.176 |    0.00217487  |        0.001  |
| getSlot                           |   57737 |      0     |      0.263 |    0.00119835  |        0.001  |
| getBlocks                         |   23778 |      0     |      0.052 |    0.000766591 |        0.001  |
| getBlock                          |   15796 |      0     |      0.125 |    0.00191523  |        0.002  |
| getMultipleAccounts               |   14288 |      0     |      0.091 |    0.00262507  |        0.001  |
| getTransaction                    |    8060 |      0     |      0.35  |    0.0124865   |        0.005  |
| getBlockTime                      |    6555 |      0     |      0.126 |    0.000809916 |        0      |
| sendTransaction                   |    5238 |      0     |      0.135 |    0.00913555  |        0.005  |
| getLatestBlockhash                |    5056 |      0     |      0.176 |    0.00319699  |        0.001  |
| getBalance                        |    1849 |      0     |      0.059 |    0.00245646  |        0.001  |
| getSignatureStatuses              |    1151 |      0     |      0.127 |    0.00217376  |        0.001  |
| getVersion                        |     600 |      0     |      0.13  |    0.00216333  |        0.001  |
| isBlockhashValid                  |     557 |      0     |      0.026 |    0.000570916 |        0      |
| getSignaturesForAddress           |      64 |      0     |      0.016 |    0.00207813  |        0.001  |
| simulateTransaction               |      38 |      0     |      0.001 |    0.0005      |        0.0005 |
| requestAirdrop                    |      32 |      0     |      0.002 |    0.001       |        0.001  |
| getHealth                         |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000666667 |        0.001  |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |        0.001  |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

afalaleev commented on common/neon/cu_price_data_model.py:
```diff
@@ -0,0 +1,91 @@
+            + (end_val - start_val) * (pp - self._PercentileList[biggest_known_p_idx - 1]) / self._PercentileStep
+        )
+
+    def get_median(self) -> float:
```
Not used method

evgenyzdanovich replied:
done

afalaleev commented on common/neon/cu_price_data_model.py:
```diff
@@ -0,0 +1,91 @@
+    _PercentileCount: Final[int] = 11  # 100 / step + 1.
+    _PercentileList: Final[list[int]] = [i * 10 for i in range(11)]  # 0, 10, ..., 100
+
+    _default: ClassVar[list[int] | None] = None
```
Wrong nested type:
```
_default: ClassVar[CuPricePercentilesModel | None] = None
```

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -96,16 +97,47 @@ async def start(self) -> None:
+                AND t.block_slot <= {latest_slot}
+              GROUP BY
+                block_slot
+              LIMIT {num_blocks}
```
The nested query doesn't have an ORDER
So, it isn't clear what blocks will be returned by the request.

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -96,16 +97,47 @@ async def start(self) -> None:
+              WHERE
+                t.chain_id = {chain_id}
+                AND t.max_fee_per_gas != ''
+                AND t.block_slot <= {latest_slot}
```
We can predict the start block slot here.
Solana loses ~5% of blocks on forks.
Let's increase this number in 4 times to 20%.
So the condition can be accurate and include the start slot.

afalaleev replied:
I am worried about a living system, where there are already several tens of millions of blocks.

evgenyzdanovich replied:
done

afalaleev commented on indexer/db/neon_tx_db.py:
```diff
@@ -298,21 +350,28 @@ def to_meta(self) -> NeonTxMetaModel | None:
-        neon_tx = NeonTxModel(
+        params = dict(
             tx_type=self.tx_type,
+            tx_chain_id=None if self.chain_id == 0 else self.chain_id,
```
It looks like there can be None in self.chain_id ...
Is it the correct condition for the case of chain_id = None ?

evgenyzdanovich replied:
well, either way, if chain_id = None, it will be assigned as None, regardless if the condition is `self.chain_id == 0` or `not self.chain_id`.

But I agree, it looks weird, so I make it more "pythonic".

afalaleev commented on indexer/db/indexer_db_client.py:
```diff
@@ -94,6 +99,15 @@ async def get_finalized_block(self) -> NeonBlockHdrModel:
+        return await self._sol_block_db.get_cu_price_percentile_list(None, num_blocks, latest_block)
+
+    async def get_recent_priority_fees(self, num_blocks: int) -> list[PriorityFeePercentiles]:
+        return await self._sol_block_db.get_cu_price_percentile_list(None, num_blocks, _MAX_LATEST_BLOCK_SLOT)
```
I disagree with the condition.
You should pass `self.get_latest_slot()`

evgenyzdanovich replied:
done

afalaleev commented on proxy/base/mp_api.py:
```diff
@@ -164,6 +167,50 @@ def is_empty(self) -> bool:
+
+        idx: int = bisect_left(self.token_gas_prices, ephemeral_block_ts, key=lambda v: v.timestamp)
+        if self.token_gas_prices[idx].timestamp != ephemeral_block_ts:
+            idx -= 1
```
I'm paranoic, but it is better to check that idx >= 0

evgenyzdanovich replied:
Added the check, although I'm sure it's unnecessary.

bisect_left simply can't return negative values. it returns the non-negative index where the value should be inserted. Also, `ephemeral_block_ts` can't be less than `earliest_ts`, because of the way it's calculated.

=> `bisect_left` in our case returns 0 if and only if the `earliest_ts` is equal to the `ephemeral_block_ts` (hence the check).

![image](https://github.com/user-attachments/assets/ce9b0a78-0786-4a85-aef4-b8c6ff5dea4c)

afalaleev commented on proxy/rpc/np_block_transaction_api.py:
```diff
@@ -435,27 +449,62 @@ async def get_tx_by_block_hash_idx(self, block_hash: EthBlockHashField, index: H
+        mempool_basefee_gas_prices: MpRecentGasPricesModel = await self._server.get_recent_gas_prices_list(ctx)
+        if mempool_basefee_gas_prices.token_gas_prices:
+            latest_slot: int = await self._db.get_latest_slot()
+            if block.slot > latest_slot:
```
This condition doesn't depend on `mempool_basefee_gas_prices` and doesn't use it.

evgenyzdanovich replied:
reformatted

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -1,19 +1,30 @@
+from ..base.mp_api import MpGasPriceTimestamped, MpRecentGasPricesModel, MpTokenGasPriceModel, MpGasPriceModel
+
+# Maximum number of blocks a User can query in eth_feeHistory.
+_FEE_HISTORY_MAX_NUM_BLOCKS: int = 1024
```
Final[int]

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +171,198 @@ async def get_neon_gas_price(
+            if priority_fee_percentile_list[i] >= priority_fee_percentile_list[i + 1]:
+                raise EthError(message="Invalid priority fee percentiles: should be an increasing sequence.")
+
+    def _get_gas_used_ratio(self) -> float:
```
This is a static method

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +171,198 @@ async def get_neon_gas_price(
+        # Filling in the random high number in [0.85, 1] range.
+        return 0.85 + random.random() * 0.15
+
+    def _calc_priority_fee_list(
```
This is a static method

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +171,198 @@ async def get_neon_gas_price(
+        base_fee_list.append(current_gas_price)
+        return _RpcFeeHistoryResp.from_raw(base_fee_list, gas_used_ratio_list, earliest_block_slot_resp, reward_list)
+
+    def _validate_fee_history_inputs(
```
This is a static method

evgenyzdanovich replied:
done

afalaleev commented on proxy/rpc/np_gas_price.py:
```diff
@@ -140,3 +171,198 @@ async def get_neon_gas_price(
+            # Construct entries into base_fee_data_list from in-memory mempool-based recent token gas prices.
+            for gas_price in reversed(recent_basefee_price_list):
+                # Heuristically assign the block slot.
+                ephemeral_block_slot = mempool_basefee_gas_prices.get_ephemeral_block_slot(
```
IMHO these calculations can be simplified.
Mempool has access to Solana.
So, you can get the latest slot and link it to the gas-price.
And here, you can just fill non-existent slots with the same gas-price.

evgenyzdanovich replied:
nice, agree.

afalaleev commented on proxy/executor/strategy_base.py:
```diff
@@ -255,16 +272,33 @@ async def _init_sol_tx_cfg(
+                else:
+                    # For general Dynamic Gas transactions: we take the cu_price according to what the User set in Neon tx.
+                    cu_price = max(
+                        1, int(self._ctx.max_priority_fee_per_gas * 1_000_000 * 5000.0 / (base_fee_per_gas * cu_limit))
```
EIP-1559 is about SAVING money on PRIORITY
It isn't about FORCING the PRIORITY by a user.

The rule should be
- User sets the MAX priority fee (NeonTx.priority_fee)
- Proxy calculates the CURRENT priority fee in network (network.priority_fee)
- Proxy sets the SolanaTx.priority_fee = min(NeonTx.priority_fee, network.priority_fee)
- EVM  pays to operator min(NeonTx.priority_fee, SolanaTx.priority_fee)

It means that EVM protects the User from the Operator desire to get more money on priorities.

afalaleev replied:
```
   cu_price = min(
      await self._estimate_cu_price(),
      max(
         1, 
         int(self._ctx.max_priority_fee_per_gas * 1_000_000 * 5000.0 / (base_fee_per_gas * cu_limit)
      )
   )
 ```

evgenyzdanovich replied:
done

evgenyzdanovich committed:
Review changes.

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   61436 |      0     |      0.193 |    0.0014643   |         0.001 |
| getAccountInfo                    |   59850 |      0     |      0.212 |    0.00268872  |         0.001 |
| getBlocks                         |   25227 |      0     |      0.116 |    0.00094161  |         0.001 |
| getBlock                          |   16576 |      0     |      0.057 |    0.00231708  |         0.002 |
| getMultipleAccounts               |   14552 |      0     |      0.229 |    0.00320877  |         0.001 |
| getLatestBlockhash                |   11647 |      0     |      0.142 |    0.00228188  |         0.001 |
| getTransaction                    |    8146 |      0.001 |      0.205 |    0.014348    |         0.006 |
| getBlockTime                      |    6558 |      0     |      0.109 |    0.00110689  |         0.001 |
| sendTransaction                   |    5309 |      0.001 |      0.21  |    0.010815    |         0.005 |
| getBalance                        |    1861 |      0     |      0.083 |    0.00303117  |         0.001 |
| getSignatureStatuses              |    1178 |      0     |      0.108 |    0.00247623  |         0.001 |
| getVersion                        |     614 |      0     |      0.054 |    0.00223453  |         0.001 |
| isBlockhashValid                  |     570 |      0     |      0.029 |    0.000847368 |         0     |
| getSignaturesForAddress           |      59 |      0.001 |      0.027 |    0.00272881  |         0.002 |
| simulateTransaction               |      38 |      0     |      0.002 |    0.000578947 |         0.001 |
| requestAirdrop                    |      32 |      0     |      0.002 |    0.00109375  |         0.001 |
| getHealth                         |       3 |      0     |      0.002 |    0.001       |         0.001 |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |         0     |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0.01  |      0.01  |    0.01        |         0.01  |

</details>

BoogVAr commented on proxy/base/rpc_api.py:
```diff
@@ -65,8 +71,8 @@ def to_core_tx(self, chain_id: int) -> CoreApiTxModel:
             gas_limit=self.gas,
-            gas_price=self.gasPrice,
-            chain_id=chain_id
+            gas_price=self.gasPrice if self.txType == 0 else self.maxFeePerGas - self.maxPriorityFeePerGas,
```
is it possible to have a negative value? Do we need an additional check?

evgenyzdanovich replied:
agree, added a corresponding pydantic post init validation.

BoogVAr commented on indexer/db/neon_tx_db.py:
```diff
@@ -298,21 +355,28 @@ def to_meta(self) -> NeonTxMetaModel | None:
+        # TODO EIP1559: introduce blob field which stores rlp and construct via from_raw(rlp).
+        # Alternatively, allow non-frozen model and modify it in the model_post_init.
+        NeonTxModel.pop_ctr_params(params)
+        neon_tx = NeonTxModel(**params)
```
NeonTxModel - from_raw? -> from_dict ? I suggest to add a method from_dict and check dict in from_raw.

BoogVAr replied:
`NeonTxModel.pop_ctr_params(params)`
Please put it inside the from_dict method to be sure it works every time someone use the method and don't forget about.

evgenyzdanovich replied:
agree

BoogVAr commented on indexer/db/neon_tx_db.py:
```diff
@@ -359,6 +424,12 @@ def _decode_event_list(value: bytes) -> list[NeonTxEventModel]:
+@dataclass(frozen=True)
+class BlockFeeGasData:
+    block_slot: int
+    average_base_fee: float
```
what do you think about int here? or Decimal if actually needed.

evgenyzdanovich replied:
int would suffer from rounding error and it would make "baseFeePerGas" in RPC methods a little less precise.
decimal is an overkill imo, because this field is a result of an sql AVG which has its own (limited) representation of floating point numbers.

I think decimal is more appropriate when you need to do some math on non-integer number and you want to preserve more precision.

BoogVAr replied:
Ok. Agree, thank you.

evgenyzdanovich committed:
Minor review fixes.

BoogVAr reviewed (approved):
Approved

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   59157 |      0     |      0.131 |    0.00165629  |         0     |
| getSlot                           |   57040 |      0     |      0.145 |    0.000995617 |         0     |
| getBlocks                         |   22476 |      0     |      0.054 |    0.000678813 |         0.001 |
| getBlock                          |   14518 |      0     |      0.048 |    0.00156185  |         0.001 |
| getMultipleAccounts               |   13576 |      0     |      0.117 |    0.00222253  |         0.001 |
| getLatestBlockhash                |   10804 |      0     |      0.124 |    0.00173547  |         0.001 |
| getTransaction                    |    8098 |      0     |      0.152 |    0.0106419   |         0.004 |
| getBlockTime                      |    6384 |      0     |      0.058 |    0.00060683  |         0     |
| sendTransaction                   |    5267 |      0     |      0.133 |    0.00819727  |         0.004 |
| getBalance                        |    1851 |      0     |      0.11  |    0.00167153  |         0.001 |
| getSignatureStatuses              |    1156 |      0     |      0.125 |    0.00159429  |         0     |
| getVersion                        |     602 |      0     |      0.04  |    0.00134385  |         0     |
| isBlockhashValid                  |     560 |      0     |      0.03  |    0.000557143 |         0     |
| getSignaturesForAddress           |      60 |      0     |      0.003 |    0.00146667  |         0.001 |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000421053 |         0     |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00096875  |         0.001 |
| getHealth                         |       3 |      0     |      0.001 |    0.000666667 |         0.001 |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0.051 |      0.051 |    0.051       |         0.051 |

</details>
